File: arguments.py
import os
import sys

from exception import CommandLineException


class ArgumentParser(object):
    class _ArgumentParserResults(object):
        def __init__(self, command, parameters):
            self.command = command
            self.parameters = parameters

        def __str__(self):
            return self.__repr__()

        def __repr__(self):
            return f"{self.command}:{self.parameters}"

    arguments = sys.argv[1:]

    def parse(self):
        command, parameters = None, []
        for index, argument in enumerate(self.arguments):
            if index == 0:
                command = argument
                continue
            is_valid_command = argument.startswith("--")
            if not is_valid_command:
                CommandLineException(
                    f"{argument} is not a valid parameter key",
                    "parameters should start with double hiphens",
                )
            parameters.append(argument[2:])
        return self._ArgumentParserResults(command, parameters)

File: src/arguments.ts
import { argv } from "process";
import { CommandLineException } from "./exception";

export interface ArgumentParserResults {
  // The command that the user wants to run
  // Eg := tree, run, list, help etc
  // if the length of the arguments is 0
  // The details about the current working directory
  // is displayed
  command: string | null;
  // The flags passed in along with the command
  // Eg := --only-dirs, --gitignore
  // Flags should start with double hiphens(--)
  parameters: Array<string>;
}

export class ArgumentParser {
  // The list of command line arguments
  private readonly arguments: Array<string> = argv.slice(2);

  /**
   * @oublic
   *
   * The argument parser function that
   * parses the list if command line arguments
   * into commands and parameters(or flags)
   *
   * @returns {ArgumentParserResults} The command and the flags
   */
  public parseArguments(): ArgumentParserResults {
    let command: string | null = null;
    let parameters: Array<string> = new Array<string>();
    for (let index = 0; index < this.arguments.length; index++) {
      const currentArgument: string = this.arguments[index];
      if (index == 0) {
        // If the length of the list is greater than 0
        // The first element of the list is taken as the
        // command
        command = currentArgument;
        continue;
      }
      const isValidArgument: boolean = currentArgument.startsWith("--");
      if (!isValidArgument) {
        const error = new CommandLineException({
          message: `${currentArgument} is not a valid argument`,
          suggestion: "Parameter keys should start with --",
        });
      }
      parameters.push(currentArgument.slice(2));
    }
    return { command: command, parameters: parameters };
  }
}
